stellargraph/data/edge_cache.py

```diff
@@ -0,0 +1,206 @@
+            raise ValueError("Unknown cache type: {}".format(cache_type))
+        else:
+            return type.__call__(cls, *args, **kwargs)
+
+
+class EdgeCache(metaclass=EdgeCacheFactory):
+    """
+    Encapsulation of node and edge data, for neighbourhood
+    queries.
+
+    Args:
+        cache_type (EdgeCacheType): The type of cache to use.
+        edge_data (EdgeData): The edge data object.
+        node_data (NodeData): The node data object.
+    """
+
+    def neighbour_nodes(self, node_id: Any) -> Iterable[Any]:
+        """
+        Obtains the collection of neighbouring nodes connected
+        to the given node, without regards to directionality.
+        Nodes may be repeated if there are multiple edges.
+
+        Args:
+            node_id (any): The identifier of the node in question.
+
+        Returns:
+            iterable: The identifiers of the neighbouring nodes.
+        """
+        raise NotImplementedError
+
+    def in_nodes(self, node_id: Any) -> Iterable[Any]:
+        """
+        Obtains the collection of neighbouring nodes with edges
+        directed to the given node. Nodes may be repeated if there
+        are multiple edges.
+
+        Args:
+            node_id (any): The identifier of the node in question.
+
+        Returns:
+            iterable: The identifiers of the neighbouring in-nodes.
+        """
+        raise NotImplementedError
+
+    def out_nodes(self, node_id: Any) -> Iterable[Any]:
+        """
+        Obtains the collection of neighbouring nodes with edges
+        directed from the given node. Nodes may be repeated if there
+        are multiple edges.
+
+        Args:
+            node_id (any): The identifier of the node in question.
+
+        Returns:
+            iterable: The identifiers of the neighbouring out-nodes.
+        """
+        raise NotImplementedError
+
+
+#############################################
+# Class for direct access to the edge data:
+
+
+class NoEdgeCache(EdgeCache):
+    """
+    Delegates to the underlying edge data.
+
+    Args:
+        edge_data (EdgeData): The edge data object.
+        node_data (NodeData): The node data object.
+    """
+
+    def __init__(self, edge_data: EdgeData, node_data: NodeData):
+        self._edge_data = edge_data
+        # Ensure all nodes are in the graph
+        for source_id, target_id in edge_data.edges():
+            node_data.add_node(source_id)
+            node_data.add_node(target_id)
+
+    def neighbour_nodes(self, node_id: Any) -> Iterable[Any]:
+        return self._edge_data.neighbour_nodes(node_id)
+
+    def in_nodes(self, node_id: Any) -> Iterable[Any]:
+        return self._edge_data.in_nodes(node_id)
+
+    def out_nodes(self, node_id: Any) -> Iterable[Any]:
+        return self._edge_data.out_nodes(node_id)
+
+
+#############################################
+# Class for cached access to the edge data, ignoring
+# node and edge type information.
+
+
+class UntypedEdgeCache(EdgeCache):
```

Comment:
I think this is essentially SDAL?

Reply:
Yes, I started with the simplest case of homogeneous graphs, mainly because that is where I have spent most of my time (e.g. GraphSAGE). I have really only just started looking seriously at how heterogeneous graphs are being used (e.g. HinSAGE).